### PR TITLE
add: CSP: Report only #8520

### PR DESCRIPTION
**Pull request by cnsgithub**
```xml
    <context-param>
        <param-name>primefaces.CSP_REPORT_ONLY</param-name>
        <param-value>true</param-value>
    </context-param>
    <context-param>
        <param-name>primefaces.CSP_REPORT_ENDPOINT</param-name>
        <param-value>/csp-reporting</param-value>
    </context-param>
```

grafik

### CODE REVIEW

File: docs/12_0_0/core/contentsecuritypolicy.md [hunk 1]
```diff
@@ -19,6 +19,20 @@ CSP is disabled by default and a global parameter is required to turn it on.
     <param-value>true</param-value>
 </context-param>
 ```
+
+In large web applications you might decide to report CSP violations only as a start instead of directly forcing them. Browsers will report CSP violations to a user-defined HTTP endpoint then.
```
Comment (melloware) [inline docs/12_0_0/core/contentsecuritypolicy.md:23]: Dont forget to also add to /gettingstarted/configuration.md file

File: primefaces/src/main/java/org/primefaces/csp/CspPhaseListener.java [hunk 2]
```diff
@@ -67,7 +73,20 @@ public void beforePhase(PhaseEvent event) {
         CspState state = PrimeFacesContext.getCspState(context);
 
         String policy = LangUtils.isBlank(customPolicy.get()) ? "script-src 'self'" : customPolicy.get();
-        externalContext.addResponseHeader("Content-Security-Policy", policy + " 'nonce-" + state.getNonce() + "'");
+        policy += " 'nonce-" + state.getNonce() + "';";
+
+        if (LangUtils.isNotBlank(reportEndpoint.get())) {
+            // current browser compatibility
+            policy += " report-uri " + reportEndpoint.get() + ";";
+            // TODO forward compatibility via report-to directive and Report-To header left out for the moment to reduce complexity
```
Comment (melloware) [inline primefaces/src/main/java/org/primefaces/csp/CspPhaseListener.java]: Remove TODO?

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by melloware** [comment]
@cnsgithub long time!   Thanks for the PR this looks pretty good to me just a couple of comments.

**Comment by cnsgithub** [comment]
> just a couple of comments.

Done.

Thanks for the quick review, @melloware

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
The only other thing you might want to add is `/gettingstarted/whatsnew.md` just to let people know the new options are available.

**Comment by tandraschko** [comment]
IMO we should introduce `CSP_POLICY_REPORT_ONLY` as new param for allowing setting multiple params, not just report-uri

and maybe instead of `CSP_REPORT_ONLY` `true|false`, change `CSP` to support `true|false|reportOnly` ?

**Comment by melloware** [comment]
or we could just use 1 param...

```xml
    <context-param>
        <param-name>primefaces.CSP_REPORT_ONLY</param-name>
        <param-value>/csp-reporting</param-value>
    </context-param>
```

basically if the value for REPORT_ONLY exists its considered report only and we use the value as the endpoint?

**Comment by tandraschko** [comment]
i dont like to introduce 1 param in web.xml for multiple header values (1:N), therefore i would follow `CSP_POLICY`

**Comment by melloware** [comment]
No problem.  I like your idea then of `CSP` being a value.

**Comment by cnsgithub** [comment]
> and maybe instead of `CSP_REPORT_ONLY` `true|false`, change `CSP` to support `true|false|reportOnly` ?

The disadvantage of this would be that CSP cannot be turned on together with reporting?

Where do you want to specify the `report-uri`? In both `CSP_POLICY_REPORT_ONLY ` and `CSP_POLICY`?

**Comment by tandraschko** [comment]
hmmmm, ok we have 4 possible modes now:
- CSP off (`CSP=false`)
- CSP on (`CSP=true`)
- CSP on with reporting (`CSP=true`, `CSP_POLICY="report-uri /csp-violation-report-endpoint/"`)
- CSP off but reporting only

so the 4) is the only new mode now
and i think `CSP="reportOnly"` and `CSP_REPORT_ONLY_POLICY="report-uri /csp-violation-report-endpoint/"` is fine

**Comment by tandraschko** [comment]
any update?

**Comment by christophs78** [comment]
MDN says report-uri is deprecated: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri

**Comment by cnsgithub** [comment]
> MDN says report-uri is deprecated: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri

That's true in theory but not in practice since not all browsers support the new `Report-To` header: https://caniuse.com/?search=report-to

grafik

**Comment by cnsgithub** [comment]
> any update?

Sorry, currently no time, maybe someone else wants to take over, otherwise we would have to wait for a while.

**Comment by tandraschko** [comment]
then i will close it for now

**Pull request closed without merging**